### PR TITLE
hooks: Add BeforeStep and AfterStep Hooks

### PR DESCRIPTION
**Pull request by Adam-ARK**
Another PR attempt to get BeforeStep/AfterStep hooks (https://github.com/cucumber/cucumber-js/pull/1058) promoted into the project (hopefully for the 7.0.0 release), since the current open PR(https://github.com/cucumber/cucumber-js/issues/997) hasn't been touched since May. This PR used PR https://github.com/cucumber/cucumber-js/issues/997 as the base, fixing all existing issues to get all unit/feature tests passing, and fixed merge conflicts with latest code. Then used the feature files from PR https://github.com/cucumber/cucumber-js/pull/1058 to add tests for functionality. 

Also updated files under /docs/ to describe the new hook usage. 

Not sure what all is still required before this attempted solution can be merged, besides possibly more unit tests?

### CODE REVIEW

**Comment by Adam-ARK** [comment]
@charlierudolph Could you take a look at this PR when you get the chance?

@davidjgoss Is there a possibility of getting this into the 7.0.0 release?

**Comment by mpkorstanje** [comment]
@aslakhellesoy can you review this or find some one to review it?

**Comment by aslakhellesoy** [comment]
I’ll try to get it done this week

**Comment by davidjgoss** [comment]
I should be able to review tonight - sorry for delay

File: src/runtime/pickle_runner.ts
Comment (davidjgoss) [inline src/runtime/pickle_runner.ts:440]: I think we would still want to run the `AfterStep` hooks if the step failed. A common cited use case for this feature is to take a screenshot or attach an API response if the step fails.

File: docs/support_files/api_reference.md
Comment (davidjgoss) [inline docs/support_files/api_reference.md:71]: Could we link to the current Gherkin?

File: docs/support_files/hooks.md
Comment (davidjgoss) [inline docs/support_files/hooks.md:106]: This should be `##` for an h2

File: docs/support_files/hooks.md
Comment (davidjgoss) [inline docs/support_files/hooks.md:117]: I think it would be good in this example or another to show the function consuming the step result, as users will want to know how to do this.

File: src/support_code_library_builder/types.ts
Comment (davidjgoss) [inline src/support_code_library_builder/types.ts:20]: Would we ever not provide the result?

**Review by davidjgoss — CHANGES_REQUESTED** [review]
As you mentioned we should add appropriate unit test coverage for `pickle_runner` and `support_code_library_builder`.

Also a recent addition is that we emit messages for support code definitions that can be picked up by formatters - could you follow this pattern for these new hooks? See https://github.com/cucumber/cucumber-js/blob/master/src/cli/helpers.ts#L246

**Comment by Adam-ARK** [comment]
@davidjgoss I've finished addressing all your code review comments and adding unit tests where appropriate. Please let me know of any other changes/updates I still need to make to this pull request.

**Comment by Adam-ARK** [comment]
@davidjgoss  Can you review the PR again whenever you have time? Would really like to try to get this in for the 7.0.0 full release for use.

**Review by aslakhellesoy — APPROVED** [review]
This looks really solid to me. I have no comments. @davidjgoss @charlierudolph ?

**Review by davidjgoss — APPROVED** [review]
LGTM!

**Comment by Adam-ARK** [comment]
@charlierudolph Any review comments? Or is this good to merge?

**Comment by Adam-ARK** [comment]
@davidjgoss Are we just waiting for @charlierudolph to review/approve? I've seen he's been active on other PRs in the past week but not this one. Just trying to make sure this can get in for 7.0.0 if possible.

File: src/cli/helpers.ts
Comment (aslakhellesoy) [inline src/cli/helpers.ts:226]: Would it not make more sense to emit the before step hooks before the testStepStarted messages are emitted, and the after step hooks after?

**Comment by aslakhellesoy** [comment]
We're due to release 7.0.0 in a week. I think we should merge this by then unless @charlierudolph has any reservations.

File: src/cli/helpers.ts
Comment (Adam-ARK) [reply src/cli/helpers.ts:226]: @davidjgoss @aslakhellesoy Is my assumption about what you mean correct? I can work on this today as soon as I get a confirmation (or correction if my understanding is wrong).

File: src/cli/helpers.ts
Comment (charlierudolph) [inline src/cli/helpers.ts:233]: I don't think step hooks should be emitted as cucumber-messages doesn't have a way to differentiate these from scenario hooks. Also we don't independently report a result for step hooks with them being bundled into steps. 

I think initially these should be more internal to cucumber-js. After some usage we can see if we want to keep it that way or find a way to report them (and their results) in messages.

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:474]: If we add a duration in here I think we can simplify code to avoid checking if duration is null both in this function and in the runStep function

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:462]: I think this could be restructed as

```js
const beforeStepHooksResult = /***/
let cumulatedStepResult = beforeStepHooksResult
if (beforeStepHooksResult.status !== Status.FAILED) {
  // run step
  // update cumulatedStepResult
  // run after step hooks
  // update cumulatedStepResult
}
return cumulatedStepResult;
```
I think that tightens up where variables are defined and used vs being spread across multiple if statements

File: docs/support_files/api_reference.md
Comment (charlierudolph) [inline docs/support_files/api_reference.md:98]: `After` -> `AfterStep`

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:395]: I think this should have the cumulatedTestStepResult not the result of the pickle. 

Thoughts on also having the testStepId?

File: docs/support_files/api_reference.md
Comment (charlierudolph) [inline docs/support_files/api_reference.md:70]: The argument is quite old. Can we update this with what we actually pass in?

File: docs/support_files/api_reference.md
Comment (charlierudolph) [inline docs/support_files/api_reference.md:39]: Can we also update for what this argument is since its outdated? Probably fine to just point to the type definition of what we pass in

File: src/support_code_library_builder/index_spec.ts
Comment (charlierudolph) [inline src/support_code_library_builder/index_spec.ts:259]: After -> AfterStep

File: src/support_code_library_builder/index_spec.ts
Comment (charlierudolph) [inline src/support_code_library_builder/index_spec.ts:347]: this.Before -> BeforeStep

**Comment by charlierudolph** [comment]
Thanks for contributing @Adam-ARK, sorry it took me so long to review

File: docs/support_files/api_reference.md
Comment (Adam-ARK) [reply docs/support_files/api_reference.md]: Updated.

File: docs/support_files/api_reference.md
Comment (Adam-ARK) [reply docs/support_files/api_reference.md]: Updated.

File: src/cli/helpers.ts
Comment (Adam-ARK) [reply src/cli/helpers.ts]: Removed function per Charlie's Comment about it.

File: src/runtime/pickle_runner.ts
Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts:460]: Restructured the function per suggestion.

File: src/runtime/pickle_runner.ts
Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts:397]: Not sure exactly why we would want the cumulated? The beforeStep would always get an empty result (since nothing has run yet), and the afterStep Hook would want the pickle from the actual step definition, if I'm understanding right. I did try using cumulatedResult / updating the result, but wasn't successful in my attempts. 

I did go ahead and add the stestStepId.

File: src/runtime/pickle_runner.ts
Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts:474]: What exactly would I need to set it to? I tried:
        duration: this.result?.duration !== null ? this.result.duration : 0,
but this failed unit tests.

**Comment by Adam-ARK** [comment]
@charlierudolph Address most review comments and left a few questions. 

@aslakhellesoy Any idea how to fix the cucumber reports feature file is failing in the pipelines? Seems like it's getting a 500 status error from messages.cucumber.io/api/reports

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [reply src/runtime/pickle_runner.ts:474]: In `src/time.ts` there is a method `getZeroDuration`. And then in places where you need to add durations you should no longer need null checks.

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [reply src/runtime/pickle_runner.ts:397]: Thanks for adding testStepId. 

Agree for beforeStep it won't have a result.

For afterStep I think it makes more sense to have the stepResult for the following case:
* scenario with 2 steps, step 1 fails
* if afterStep gets scenario result, both step1 and step2 afterStep get failed result
* if afterStep gets step result, step1 afterStep gets failed result, step2 afterStep gets skipped result

I think the latter is more correct as the afterStep should only care about the current step. If you care about the scenario result, you can use the after hook.

File: docs/support_files/api_reference.md
Comment (charlierudolph) [inline docs/support_files/api_reference.md:39]: Thanks for the update. This should also have `testCaseStartedId` right?

And for beforeStep doc, it should have that + `testStepId` now right?

File: src/runtime/pickle_runner.ts
Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts]: Figured out how to get this to work. Changed it so it sets stepResult to undefined at the start of each step (before running beforeStepHook so it doesn't maintain the stepResult from previous step.

**Comment by Adam-ARK** [comment]
@charlierudolph I think I've addressed all the code review comments hopefully with good resolutions/solutions.

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:56]: I disagree with having this as a class variable.  I think it can just be passed into the `runStepHook` function

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:473]: I think this should be `getZeroDuration()` (never this.result.duration). The `this.result` duration is the sum of all steps. We want to start with zero so we only sum up the duration of this set of step hooks.

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:490]: This null check should be able to be removed

File: src/runtime/pickle_runner.ts
Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts]: Updated to pass directly into the function

**Comment by Adam-ARK** [comment]
@charlierudolph Is my solution to passing the stepResult to the hook functions good? Any other comments?

**Comment by Adam-ARK** [comment]
@charlierudolph @aslakhellesoy Is there anything left for me to do to get this PR merged? I know you said you wanted to release 7.0.0 soon, so want to make sure I'm being active in addressing anything so this gets in for the release / doesn't delay it a ton.

**Comment by charlierudolph** [comment]
Hey @Adam-ARK, so I fixed the build failure if you pull in latest. And also did a bit of a refactor that will removed a function you were using (as it was an outdated method of doing things)

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:434]: Okay think this needs to be re-structured to something like:
```
// At this point lets run all stepHooks + step
const beforeStepHookResults = await this.runStepHooks(/* */); // hooks do not receive "result" in parameter
const const stepResult = await this.invokeStep(/* */);
const afterStepHookResults = await this.runStepHooks(/* */); // hooks receive stepResult as "result" in parameter (not cumulatedResult)
// Then accumulate result, use query to find the worst result and sum the durations into it
```

Sorry for the rework here. Just trying to figure out what is both simple and easy to explain to users.

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:468]: Lets return the array of results here instead of a single result, shouldn't need any accumulation

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:392]: This isSkippingSteps check should go away. We'll run the step hook if we are running the step. This function could probably be removed given the other simplifications and merge this into runStepHooks

**Comment by Adam-ARK** [comment]
@charlierudolph Apologies for the bad commit history, accidentally did a merge instead of rebase with master. 

Got everything working again with your changes, and did the changes you requested. Now there's just 1 runStepHooks function, and removed isSkippingStep(), and returns the array

For the runStep function, I still have it not run the step itself if any of the beforeStep hooks fail, as this is in-line with how the Before Scenario hooks won't run a scenario if they fail, unless we want them to be different?

Otherwise I refactored the function with this workflow:
// Run before hooks, step and after hooks
// Set the final result to be the stepDef result since it has the messages data
// get the worst status from the step/hooks and update the final result with it
// sum all the durations together into the final result

**Comment by Adam-ARK** [comment]
@charlierudolph  @davidjgoss Any other comments / review for this PR? Or are we good to finally merge this?

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:415]: Can we rename this from `beforeStepHooksResult` to `beforeStepHookResults`? I think moving the `s` to the end make it easier to tell this is an array.

Same for `afterStepHooksResult` -> `afterStepHookResults`.

**Comment by charlierudolph** [comment]
I agree with not running the step if a beforeStepHook fails but we should always run afterStepHooks.

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:442]: This looks like we only sum the durations if all beforeHooks passed and we should sum it no matter what.

**Comment by Adam-ARK** [comment]
@charlierudolph Updated the variable names, and made the change so that the only thing that won't happen if the beforeStep Hooks fail, is that the step itself won't run, but afterStep Hooks / sum duration will.

File: src/runtime/pickle_runner.ts
Comment (charlierudolph) [inline src/runtime/pickle_runner.ts:434]: I think we need to assign more than just status here (things like message are also important). And I think the duration below should be recalculated from scratch (if a before hook fails I believe its duration would be counted twice)

Thoughts on 
```ts
const stepResults = /* run before step hook */;
// if worst step test result status is not failed, run step and append to stepResults
// concat after step hook results to step results
const finalStepResult = /* worst step test result */;
finalStepResult.duration = getZeroDuration();
// iterate though step results, incrementing duration
return finalStepResult;
```

Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts]: Updated to address this. There's two weird things I found while doing this:

1) If I concatenate the afterStepHook Results into the stepResults, then if an after step hook failed it wasn't getting picked up. I could only get it to pickup if I manually in a for loop pushed each individual result into the stepResults array. Not sure why this is, but this was the only way I could get it to work.

2) Had to make a separate variable to sum durations into and then assign that to the finalStepResult. If I just did what you suggested, I kept running into either getting twice the duration, or always getting zero duration. I tried a lot of different ways of doing it, but what I have is the only one I could actually get to work properly.

Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts]: @davidjgoss  Would you also have any idea as to my questions on this thread?

Comment (davidjgoss) [reply src/runtime/pickle_runner.ts]: Sorry for late reply - I should be able to have a look at this today/tonight.

Comment (charlierudolph) [reply src/runtime/pickle_runner.ts]: For concat, did you reassign the variable? (concat returns the new array instead of altering the variable)

For the sum durations, current implementation looks fine.

Comment (charlierudolph) [reply src/runtime/pickle_runner.ts]: Regarding the durations, can you please add a unit test for that?

Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts]: Yup, forgot to reassign the variable for concat, not sure how I missed that.

Also added a unit test with the other duration ones in summary_helper_spec.ts

Comment (Adam-ARK) [reply src/runtime/pickle_runner.ts]: @charlierudolph Are we finally good to merge this PR?

Comment (charlierudolph) [reply src/runtime/pickle_runner.ts]: Yep. Looks good! Thanks for your contribution!

**Comment by aslakhellesoy** [comment]
Hi @Adam-ARK,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Pull request merged**
